README.md
<h1 align="center">
  <br>
  snip-lookup.nvim
  <br>
  <br>

![aoeu](https://user-images.githubusercontent.com/47462344/204839108-6dc32a57-1c4b-4921-911e-5220de4a7de8.gif)

</h1>
<h2 align="center">
  <img alt="PR" src="https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat"/>
  <img alt="Lua" src="https://img.shields.io/badge/lua-%232C2D72.svg?&style=flat&logo=lua&logoColor=white"/>
  <img alt="Rust" src="https://img.shields.io/badge/-Rust-orange"/>
</h2>
<h3 align="center">
  <a href="#requirements">Requirements</a> •
  <a href="#installation-and-setup">Installation and Setup</a> •
  <a href="#usage">Usage</a> •
  <a href="#documentation">Documentation</a>
</h3>

Have you amassed such a vast collection of snippets, that remembering the prefixes to use each one is nigh-impossible?

Do you find yourself with too many things you _want_ as snippets, but not enough time to _create all of them_?

**Enter: snip-lookup.nvim!**

---

Let me preface with saying that I LOVE snippets. Whether I'm using them to
quickly type out my home address, or for commonly-typed code-snippets - I find
myself constantly using or creating them.

That being said, I also have have several "wants" that I (until now) _haven't_ been able to fulfill:

- I want editing my snippets to be fast and easy, so I can tweak them on the fly
- I want to version control my snippets, so I can use them across multiple computers
- I want to categorize my snippets, so I can quickly know what snippets I have
- I want to fuzzy-search through my snippets, so I can select snippets faster

**snip-lookup.nvim solves those by:**

- Storing all your snippets in a simple YAML file of your choosing
  - Specify a file path to easily version control your snippets!
  - For example, you can store your work-specific snippets in a work repository
- Providing a keybind for instantly editing your snippets
- Utilizing [telescope](https://github.com/nvim-telescope/telescope.nvim) to fuzzy-search through existing categories/snippets

## Requirements

- [telescope.nvim](https://github.com/nvim-telescope/telescope.nvim)
- **[Optional]** cargo and rust toolchain

## Installation and Setup

Using [packer](https://github.com/wbthomason/packer.nvim):

```lua
use {
    'rnprest/snip-lookup.nvim',
        requires = { 'nvim-telescope/telescope.nvim' },
        run = './install.sh',
        config = function()
            require('snip-lookup').setup {
                -- Default options are listed below - you can just call setup() if these are fine with you
                open_picker = '<leader>sl',
                open_config = '<leader>esl',
                config_file = vim.fn.stdpath 'data' .. '/snip-lookup/snippets.yaml',
            }
    end,
}
```

Build from source (requires cargo) - just tweak the 'run' arg

```lua
use {
    'rnprest/snip-lookup.nvim',
        requires = { 'nvim-telescope/telescope.nvim' },
        run = './install.sh build',
        config = function()
            require('snip-lookup').setup {
                -- Default options are listed below - you can just call setup() if these are fine with you
                open_picker = '<leader>sl',
                open_config = '<leader>esl',
                config_file = vim.fn.stdpath 'data' .. '/snip-lookup/snippets.yaml',
            }
    end,
}
```

## Usage

Searching through your snippets

1. Hit `<leader>sl`
2. Start typing the name of your category
   - Hit `<CR>` to select a category
3. Start typing the name of your snippet
   - Hit `<CR>` to copy the selected snippet to your clipboard

Editing your snippets

1. Hit `<leader>esl`
2. Start editing to your heart's content
   - To save your changes, just save the file.
   - Any changes made will be immediately available the next time you hit your search remap

## Documentation

Snippet file structure:

```
categories:
  <snippet category>:
    icon: <icon> # Setting this is OPTIONAL, and will default to blank space
    snippets:
      - <snippet name>: <snippet contents>
      - ...
  ...
```

Example (This will be the default content of your snippet file):

```yaml
categories:
  Email Addresses:
    icon: 📧
    snippets:
      - John Doe: john.doe@gmail.com
      - Jane Doe: jane.doe@gmail.com
  Phone Numbers:
    icon: 📞
    snippets:
      - Jack Black: (111) 111-1111
      - Jill Dill: (222) 222-2222
  File Templates:
    snippets:
      - README: |-
          # Title

          ## Installation

          ## Usage
  Email Groups:
    snippets:
      - Family: >-
          mom@gmail.com;
          dad@hotmail.com;
          brother@aol.com;
          sister@yahoo.com;
          son@proton.me;
          daughter@outlook.com
```

## Known Issues

If you find that your icons and category names don't share the same consistent width when searching through categories, it could be because of the emoji you're trying to use.

I'm not sure how to fix this other than by replacing the problematic emoji (demonstrated below), but if someone thinks they could find a proper solution then that would be greatly appreciated!

https://user-images.githubusercontent.com/47462344/211921679-a8b89a7f-966c-4ce0-ac36-b6e5adb5042a.mp4


CHANGELOG.md
# Changelog

## [0.3.0](https://github.com/rnprest/snip-lookup.nvim/compare/v0.2.2...v0.3.0) (2023-01-11)


### Features

* **ci:** add 2 more OS artifacts ([c47b748](https://github.com/rnprest/snip-lookup.nvim/commit/c47b7486875773f2c9f3be35c47462be71ae6608))

## [0.2.2](https://github.com/rnprest/snip-lookup.nvim/compare/v0.2.1...v0.2.2) (2023-01-11)


### Bug Fixes

* **ci:** zip release asset before uploading ([952214f](https://github.com/rnprest/snip-lookup.nvim/commit/952214ff055e273f57db5ba5777fcf594a730625))

## [0.2.1](https://github.com/rnprest/snip-lookup.nvim/compare/v0.2.0...v0.2.1) (2023-01-11)


### Bug Fixes

* try different release action ([9a0842b](https://github.com/rnprest/snip-lookup.nvim/commit/9a0842b934ea18b8f1f29308e0dd4f4b911fecf9))

## [0.2.0](https://github.com/rnprest/snip-lookup.nvim/compare/v0.1.1...v0.2.0) (2023-01-11)


### Features

* use release-please ([d5463ac](https://github.com/rnprest/snip-lookup.nvim/commit/d5463accc152d825771bc2509377ce2ac06c0f24))


### Bug Fixes

* **ci:** create initial file for pandoc action ([b02082a](https://github.com/rnprest/snip-lookup.nvim/commit/b02082aba2373f258af2b3b93937df4aa58f6dc0))
* **ci:** only attach assets once release is created ([e85bd54](https://github.com/rnprest/snip-lookup.nvim/commit/e85bd54a8a900b09596656294647b64f927b61e0))
* **ci:** pass upload_url from release job ([b4669ea](https://github.com/rnprest/snip-lookup.nvim/commit/b4669ea672c56e31332a4c7490229f3439009088))


install.sh
#!/usr/bin/env bash

# ---------------------------------------------------------------------
# Taken from https://github.com/krivahtoo/silicon.nvim;
# You should absolutely check out their plugin if you often
# find yourself taking screenshots of your code to share!
# I use it all the time.

# If you do want to use their plugin, my configuration for it is below:

# use {
# 	'krivahtoo/silicon.nvim',
# 	run = './install.sh',
# 	config = function()
# 		require('silicon').setup {
# 			font = 'Iosevka Nerd Font Mono=20',
# 			background = '#ffffff',
# 			line_number = true,
# 			shadow = {
# 				blur_radius = 7.0,
# 			},
# 			pad_horiz = 20,
# 			pad_vert = 20,
# 		}
# 		vim.keymap.set('v', '<leader>ss', [[:Silicon<CR>]])
# 	end,
# }
# ---------------------------------------------------------------------

set -e

# get current version from Cargo.toml
get_version() {
	cat Cargo.toml | grep '^version =' | sed -E 's/.*"([^"]+)".*/\1/'
}

# compile from source
build() {
	echo "Building snip-lookup.nvim from source..."

	cargo build --release --target-dir ./target

	# Place the compiled library where Neovim can find it.
	mkdir -p lua

	if [ "$(uname)" == "Darwin" ]; then
		mv target/release/libsnip_lookup.dylib lua/snip_lookup.so
	elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
		mv target/release/libsnip_lookup.so lua/snip_lookup.so
	elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW64_NT" ]; then
		mv target/release/snip_lookup.dll lua/snip_lookup.dll
	fi
}

# download the snip-lookup.nvim (of the specified version) from Releases
download() {
	echo "Downloading snip-lookup.nvim library: $1"
	if [ "$(uname)" == "Darwin" ]; then
		arch_name="$(uname -m)"
		curl -fsSL https://github.com/rnprest/snip-lookup.nvim/releases/download/$1/snip-lookup-mac-${arch_name}.tar.gz | tar -xz
	elif [ "$(expr substr $(uname -s) 1 5)" == "Linux" ]; then
		curl -fsSL https://github.com/rnprest/snip-lookup.nvim/releases/download/$1/snip-lookup-linux.tar.gz | tar -xz
	elif [ "$(expr substr $(uname -s) 1 10)" == "MINGW64_NT" ]; then
		echo "Windows build is not available yet."
		build
	fi
}

case "$1" in
build)
	build
	;;
*)
	version="v$(get_version)"
	download "$version"

	;;
esac


src/lib.rs
// lib.rs
use nvim_oxi as oxi;
use oxi::conversion::{self, FromObject, ToObject};
use oxi::serde::{Deserializer, Serializer};
use oxi::{lua, Dictionary, Function, Object};
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, path::Path};

// I can map structs into lua objects perfectly fine, but hashmaps? Absolutely not.
// Thus, this wrapper struct was born.
// NOTE: I tried to have 2 separate structs (with one being a HashMap of <i32, String>), and the
// Dictionary::from_iter at the bottom kept complaining about expecting whatever struct was passed
// in first... Using the same wrapper struct for both hashmaps was the only way I could get around
// this error

#[derive(Debug, Serialize, Deserialize)]
struct NeovimWrapper {
    pub contents: HashMap<String, String>,
}

//------------------------------------------------------------------//
//           I'm yoinking these impl's straight from the            //
//                          example here:                           //
//        https://github.com/noib3/nvim-oxi/blob/master/exam        //
//                         ples/mechanic.rs                         //
//------------------------------------------------------------------//

impl FromObject for NeovimWrapper {
    fn from_object(obj: Object) -> Result<Self, conversion::Error> {
        Self::deserialize(Deserializer::new(obj)).map_err(Into::into)
    }
}

impl ToObject for NeovimWrapper {
    fn to_object(self) -> Result<Object, conversion::Error> {
        self.serialize(Serializer::new()).map_err(Into::into)
    }
}

impl lua::Poppable for NeovimWrapper {
    unsafe fn pop(lstate: *mut lua::ffi::lua_State) -> Result<Self, lua::Error> {
        let obj = Object::pop(lstate)?;
        Self::from_object(obj).map_err(lua::Error::pop_error_from_err::<Self, _>)
    }
}

impl lua::Pushable for NeovimWrapper {
    unsafe fn push(self, lstate: *mut lua::ffi::lua_State) -> Result<std::ffi::c_int, lua::Error> {
        self.to_object()
            .map_err(lua::Error::push_error_from_err::<Self, _>)?
            .push(lstate)
    }
}

//------------------------------------------------------------------//
//              Used for deserializing our yaml config              //
//------------------------------------------------------------------//

#[derive(Debug, Serialize, Deserialize)]
struct Category {
    pub icon: Option<String>,
    pub snippets: Vec<HashMap<String, String>>,
}

#[derive(Debug, Serialize, Deserialize)]
struct SnippetConfig {
    pub categories: HashMap<String, Category>,
}

impl SnippetConfig {
    fn load<P: AsRef<Path>>(p: P) -> Result<SnippetConfig, serde_yaml::Error> {
        let data = std::fs::read_to_string(p).ok().unwrap();
        let config: SnippetConfig = serde_yaml::from_str(&data).unwrap();

        Ok(config)
    }
}

#[oxi::module]
fn snip_lookup() -> oxi::Result<Dictionary> {
    let get_categories = Function::from_fn::<_, oxi::Error>(move |path: String| {
        let mut category_names: HashMap<String, String> = HashMap::new();

        let sc = SnippetConfig::load(path).unwrap();
        let mut longest_icon_length = 0; // Used for padding each icon to the same width
        for (category_name, category_contents) in sc.categories.into_iter() {
            let icon_length: usize;

            // If an icon isn't defined for a category, use an empty string instead
            let icon = if category_contents.icon.is_some() {
                category_contents.icon.unwrap().to_string()
            } else {
                "".to_string()
            };

            icon_length = icon.chars().count();

            if icon_length > longest_icon_length {
                longest_icon_length = icon_length;
            }

            category_names.insert(category_name.to_string(), icon);
        }

        // Pad each icon so that category names all begin at same column
        for icon in category_names.values_mut() {
            if icon.chars().all(|char| char == ' ') {
                *icon = format!("{:width$}", *icon, width = longest_icon_length + 1);
            } else {
                *icon = format!("{:width$}", *icon, width = longest_icon_length);
            }
        }

        let nvim_categories = NeovimWrapper {
            contents: category_names,
        };
        Ok(nvim_categories)
    });

    // TODO: figure out a way to pass 2 closures to this function..
    // Currently, I'm having to combine both arguments into a comma-separated string
    let get_snippets = Function::from_fn::<_, oxi::Error>(move |path_and_category: String| {
        let args: Vec<&str> = path_and_category.split(",").collect();
        let path = args[0].to_string();
        let category = args[1].to_string();

        let mut snippets: HashMap<String, String> = HashMap::new();

        let sc = SnippetConfig::load(path).unwrap();
        if let Some(category) = sc.categories.get(&category) {
            for snippet in &category.snippets {
                for (name, value) in snippet.iter() {
                    snippets.insert(name.to_string(), value.to_string());
                }
            }
        }

        let nvim_snippets = NeovimWrapper { contents: snippets };
        Ok(nvim_snippets)
    });

    Ok(Dictionary::from_iter([
        ("get_categories", get_categories),
        ("get_snippets", get_snippets),
    ]))
}


lua/snip-lookup/init.lua
---@mod snip-lookup

local uv = vim.loop
-- Telescope
local pickers = require 'telescope.pickers'
local finders = require 'telescope.finders'
local conf = require('telescope.config').values
local actions = require 'telescope.actions'
local action_state = require 'telescope.actions.state'
-- Rust
local rust = require 'snip_lookup'
-- snip-lookup stuff
local config = require 'snip-lookup.config'

local M = {}

--- Loads the user's specified configuration and creates their desired keymapping
---@param opts table User-provided options, to override the default options with
M.setup = function(opts)
    -- TODO: need to error if user tries to pass an option that doesn't exist
    if opts ~= nil then
        config.opts = vim.tbl_deep_extend('keep', opts, config.default_opts)
    else
        config.opts = config.default_opts
    end

    vim.api.nvim_set_keymap(
        'n',
        config.opts['open_picker'],
        [[:lua require'snip-lookup'.search()<CR>]],
        { noremap = true }
    )

    ----------------------------------------------------------------------
    --        On first load, need to create initial config file         --
    ----------------------------------------------------------------------
    local cleaned_path = vim.fn.fnamemodify(config.opts['config_file'], ':p')
    if vim.fn.filereadable(cleaned_path) == 0 then
        print 'snip-lookup: Config directory not found - Creating'
        -- Create initial config file
        M.create_config_file(config.opts['config_file'])
    end

    -- Create Commands / Keymaps
    local command = 'e ' .. config.opts['config_file']
    vim.api.nvim_create_user_command('SnipLookupEdit', command, {})
    vim.api.nvim_set_keymap('n', config.opts['open_config'], ':SnipLookupEdit<CR>', { noremap = true })
end

--- Opens the snippet telescope picker
M.search = function()
    M._categories(require('telescope.themes').get_dropdown {})
end

--- Creates the snippets-specific telescope picker for a given category
---@param opts table User-provided options, to override the default options with
---@param prompt string The title of the telescope picker
M._snippets = function(opts, prompt)
    local snippets_results = {}

    ----------------------------------------------------------------------
    --           Load <category's> snippets from config file            --
    ----------------------------------------------------------------------
    local cleaned_path = vim.fn.fnamemodify(config.opts['config_file'], ':p')
    local path_and_categories = cleaned_path .. ',' .. prompt
    local snips = rust.get_snippets(path_and_categories)
    snips = snips.contents

    ----------------------------------------------------------------------
    --            Create telescope results table from loaded            --
    --                             snippets                             --
    ----------------------------------------------------------------------
    local index = 1
    for name, value in pairs(snips) do
        snippets_results[index] = { name, value }
        index = index + 1
    end

    ----------------------------------------------------------------------
    --                     Create telescope picker                      --
    ----------------------------------------------------------------------
    opts = opts or {}
    pickers.new(opts, {
        prompt_title = prompt,
        finder = finders.new_table {
            results = snippets_results,
            entry_maker = function(entry)
                return {
                    value = entry,
                    display = entry[1],
                    ordinal = entry[1],
                }
            end,
        },
        sorter = conf.generic_sorter(opts),
        attach_mappings = function(prompt_bufnr, map)
            actions.select_default:replace(function()
                actions.close(prompt_bufnr)
                local selection = action_state.get_selected_entry()
                local snippet = selection.value[2]
                vim.fn.setreg(vim.v.register, snippet) -- copy to clipboard (system if vim.opt.clipboard is set as such)
                vim.notify(snippet .. ' has been copied!')
            end)
            return true
        end,
    }):find()
end

--- Creates the telescope picker of the user's snippet categories
---@param opts table User-provided options, to override the default options with
M._categories = function(opts)
    local category_results = {}

    ----------------------------------------------------------------------
    --           Load <category's> snippets from config file            --
    ----------------------------------------------------------------------
    local cleaned_path = vim.fn.fnamemodify(config.opts['config_file'], ':p')
    local names = rust.get_categories(cleaned_path)
    names = names.contents

    ----------------------------------------------------------------------
    --            Create telescope results table from loaded            --
    --                             snippets                             --
    ----------------------------------------------------------------------
    local index = 1
    for category, icon in pairs(names) do
        category_results[index] = { category, icon }
        index = index + 1
    end

    ----------------------------------------------------------------------
    --                     Create telescope picker                      --
    ----------------------------------------------------------------------
    opts = opts or {}
    local prompt = 'Snippet Categories'
    pickers.new(opts, {
        prompt_title = prompt,
        finder = finders.new_table {
            results = category_results,
            entry_maker = function(entry)
                return {
                    -- entry[2] = category icon
                    -- entry[1] = category name
                    value = entry,
                    display = entry[2] .. ' ' .. entry[1],
                    ordinal = entry[1],
                }
            end,
        },
        sorter = conf.generic_sorter(opts),
        attach_mappings = function(prompt_bufnr, map)
            actions.select_default:replace(function()
                actions.close(prompt_bufnr)
                local selection = action_state.get_selected_entry()
                prompt = selection.value[1]
                M._snippets(require('telescope.themes').get_dropdown {}, prompt)
            end)
            return true
        end,
    }):find()
end

--- Creates the initial configuration file at a specified path.
--- This function should take care of creating any directories
--- that don't exist, as well as the initial file creation itself.
---@param path string The filesystem path of where the user's snippet config file should be created
M.create_config_file = function(path)
    local contents = [[
# YAML Symbol crash-course:
# '|' will respect newlines - which makes it perfect for file templates!
# Likewise, '>' will NOT respect newlines - I use this for organizing a long string
# The hyphen ('-') will avoid having a trailing linebreak
categories:
  Email Addresses:
    icon: 📧
    snippets:
      - John Doe: john.doe@gmail.com
      - Jane Doe: jane.doe@gmail.com
  Phone Numbers:
    icon: 📞
    snippets:
      - Jack Black: (111) 111-1111
      - Jill Dill: (222) 222-2222
  File Templates:
    snippets:
      - README: |-
          # Title

          ## Installation

          ## Usage
  Email Groups:
    snippets:
      - Family: >-
          mom@gmail.com;
          dad@hotmail.com;
          brother@aol.com;
          sister@yahoo.com;
          son@proton.me;
          daughter@outlook.com
        ]]

    -- Create config directory if it doesn't exist already
    local cleaned_path = vim.fn.fnamemodify(path, ':p')
    local cleaned_path_parent_dir = vim.fn.fnamemodify(path, ':p:h')
    vim.fn.mkdir(cleaned_path_parent_dir, 'p')
    -- Create initial config file, that can be tweaked by user
    uv.fs_open(cleaned_path, 'w', 438, function(err, fd)
        uv.fs_write(fd, contents, 0)
        uv.fs_close(fd)
    end)
end

return M
